File: session2/BasicLanguageConstructs.hs
-- BasicLanguageConstructs.hs
-- CECS 342
-- Haskell vs C
--
-- BasicLanguage Constructs
module BasicLanguageConstructs where

-- Variables
-- Haskell uses algorithm called type inference to infers all types
seven = 7 -- Num p => p 
-- seven = 8 error, variables in haskell can't be changed. Immutable
sq123 = let x = 123 in x * x -- defined variable x with local scope 

sq321 = x * x where x = 321

-- Functions
-- Syntatic sugar is syntax designed to make things easier to read or to express but can defined in terms of constructs already in the language
twice1 n = 2 * n -- syntatic sugar, sweetened
twice2 = \n -> 2 * n -- lambda-expression, unsweetened

twice3 = (2*) -- section, partial application on infix operators
twice4 = (*) 2 -- unsugared

average1 :: Float -> Float -> Float
average1 a b = (a + b) / 2
average2 a = \b -> (a + b) / 2
average3 = \a -> \b -> (a + b) / 2
average4 = \a b -> (a + b) / 2

h = average1 2

average5 :: (Float, Float) -> Float
average5(a, b) = (a + b) / 2

-- Conditionals
signum1 n = if n < 0 then -1
            else if n > 0 then 1
            else               0

-- Piece-wise functions with guards
signum2 n | n < 0 = -1
          | n > 0 = 1
	  | otherwise = 0

-- Recursion

-- Factorial with if-then-else
-- non-tail recursive
factorial1 n = if n == 0 then 1 else n * factorial1 (n - 1)

-- Factorial with pattern matching
-- non-tail recursive
factorial2 0 = 1
factorial2 n = n * factorial2 (n - 1)

-- Factorial with accumulating parameter
-- Tail recursive version
factorial3 n = fact_acc n 1
	where
		fast_acc 0 f = f
                fact_acc n f = fact_acc (n - 1) (n * f)
-- Factorial with a list of a range of numbers
factorial4 n = product [1..n]

-- Naive fibonacci numbers
-- T(n) = T(n - 1) + T(n - 2)
-- T(n) = O(2**n)
fib1 0 = 0
fib1 1 = 1
fib1 n = fib1(n - 1) + fib1(n - 2)

-- Fibonacci numbers with two accumulating parameters
-- T(n) = O(n)
fib2 n = fib_acc n 0 1
	where
		fib_acc 0 a _ = a
		fib_acc n a b = fib_acc (n - 1) b (a + b)

File: session2/BasicLanguageConstructs.c
// BasicLanguageConstructs.c
//
// CECS 342
//
// C vs Haskell
//
// To contrast C and Haskell see: BasicLangaugeConstructs.hs

#include <stdio.h>

// Variables
int seven = 7;

// Functions
int twice(int n){
	return 2 * n;
}

int average(int a, int b){
	++seven; // unexpected side effect
	return (a + b) / 2;
}

int signum0(int n){
	if(n < 0) return -1;
	if(n > 0) return 1;
	return 0;
}

int signum1(int n){
	if(n < 0){
		return -1;
	} else {
		if(n > 0){
			return 1;
		} else{
			return 0;
		}
	}
}

int signum2(int n){
	return (n < 0) ? -1 : ((n > 0) ? 1 : 0);
}

// Loops and recursion
int factorial1(int n){ // loop
	int f = 1;

	for(int k = 1;k <= n; ++k){
		f *= k;
	}
	return f;
}

int factorial2(int n){ // Recursion
	if(n == 0) return 1;
	else return n * factorial2(n - 1);
}

int fact_acc(int n,int f){ // recursion with accumulator
	if(n == 0) return f;
	return fact_acc(n - 1, n * f);
}

int factorial3(int n){
	return fact_acc(n, 1);
}

int fact_acc_tail(int n, int f){ // Tail recursion
start:
	if (n == 0) return f;


	f *= n;
	--n;
	goto start; // return fact_acc_tail(n - 1, n * f);
}

int factorial4(int n){
	return fact_acc_tail(n, 1);
}

// Loop
int fib1(int n){
	int a = 0;
	int b = 1;

	for(int k = 1; k <= n; ++k){
	     // (a, b) <- (b, a + b)
	     int h = a;
	     a = b;
	     b += h;
	}
	return a;
}

// Recursion with accumulating parameters
int fib_acc(int n, int a, int b){
	if (n == 0) return a;
	return fib_acc(n - 1, b, a + b);
}

int fib2(int n) {
	return fib_acc(n, 0, 1);
}

// Short-circuit evaluation (compared with Haskell's lazy evalution)
int or(int a, int b){return a || b;}
int and(int a, int b){return a && b;}

// A simple test suite
int testBasicLanguageConstructs(){
	puts("Test testBasicLanguageConstructs\n");
	printf("factorial1(5) = %d\n", factorial1(5));
        printf("factorial2(5) = %d\n", factorial2(5));
	printf("factorial3(5) = %d\n", factorial3(5));
	printf("fib1(7) = %d\n", fib1(7));
	printf("fib2(7) = %d\n", fib1(7));

	puts("");

	return 0;
}
